torch_dftd/nn/base_dftd_module.py

```diff
@@ -145,19 +138,38 @@ def calc_energy_and_forces(
                 results_list[i]["forces"] = forces[batch == i].cpu().numpy()
 
         if cell is not None:
-
             # stress = torch.mm(cell_grad, cell.T) / cell_volume
             # Get stress in Voigt notation (xx, yy, zz, yz, xz, xy)
+            assert isinstance(shift, Tensor)
+            voigt_left = [0, 1, 2, 1, 2, 0]
+            voigt_right = [0, 1, 2, 2, 0, 1]
             if batch is None:
                 cell_volume = torch.det(cell).abs()
-                stress = torch.mm(cell.grad, cell.T) / cell_volume
-                stress = stress.view(-1)[[0, 4, 8, 5, 2, 1]]
+                cell_grad = torch.sum(
+                    (pos[:, voigt_left] * pos.grad[:, voigt_right]).to(torch.float64), dim=0
+                )
+                cell_grad += torch.sum(
+                    (shift[:, voigt_left] * shift.grad[:, voigt_right]).to(torch.float64), dim=0
+                )
+                stress = cell_grad.to(cell.dtype) / cell_volume
                 results_list[0]["stress"] = stress.detach().cpu().numpy()
             else:
-                cell_volume = torch.det(cell).abs()
+                assert isinstance(batch_edge, Tensor)
                 # cell (bs, 3, 3)
-                stress = torch.bmm(cell.grad, cell.permute(0, 2, 1)) / cell_volume[:, None, None]
-                stress = stress.view(-1, 9)[:, [0, 4, 8, 5, 2, 1]].detach().cpu().numpy()
+                cell_volume = torch.det(cell).abs()
+                cell_grad = pos.new_zeros((n_graphs, 6), dtype=torch.float64)
+                cell_grad.scatter_add_(
+                    0,
+                    batch.view(batch.size()[0], 1).expand(batch.size()[0], 6),
+                    (pos[:, voigt_left] * pos.grad[:, voigt_right]).to(torch.float64),
+                )
+                cell_grad.scatter_add_(
+                    0,
+                    batch_edge.view(batch_edge.size()[0], 1).expand(batch_edge.size()[0], 6),
+                    (shift[:, voigt_left] * shift.grad[:, voigt_right]).to(torch.float64),
+                )
```

Comment:
Those calculations are almost same. The only difference is the use of scatter_add instead of the summation.
This implementation can eliminate `call[batch]` and `cell[batch_edge]`, which will cause slowdown.